### PR TITLE
Feature handling books

### PR DESCRIPTION
**Pull request by artemklishch**
BREAKING CHANGES:

*Added logic to controller, dto, repository, mapper and service

Task: https://mate.academy/learn/java-spring-boot/java-spring-boot-web-dto/linear?learnItemsFilter=All&section=practice&videoId=5174&theoryId=2800&testTaskSlug=java_spring_boot_web

### CODE REVIEW

File: src/main/java/mate/academy/dtoapp/controller/BookController.java [hunk 1]
```diff
@@ -1,11 +1,35 @@
...
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
 import org.springframework.web.bind.annotation.RequestMapping;
 import org.springframework.web.bind.annotation.RestController;
 
 @RequiredArgsConstructor
 @RestController
 @RequestMapping(value = "/books")
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/dtoapp/controller/BookController.java]:
```suggestion
@RequestMapping("/books")
```

File: src/main/java/mate/academy/dtoapp/model/Book.java [hunk 2]
```diff
@@ -1,4 +1,30 @@
...
 
+import jakarta.persistence.Entity;
+import jakarta.persistence.GeneratedValue;
+import jakarta.persistence.GenerationType;
+import jakarta.persistence.Id;
+import jakarta.persistence.Table;
+import java.math.BigDecimal;
+import lombok.Data;
+
+@Data
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/dtoapp/model/Book.java]:
```suggestion
@Getter
@Setter
```

File: src/main/java/mate/academy/dtoapp/model/Book.java [hunk 3]
```diff
@@ -1,4 +1,30 @@
...
+@Entity
+@Table(name = "books")
 public class Book {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    private String title;
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/dtoapp/model/Book.java:23]: Add required constraints

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Test your code in Postman

File: src/main/java/mate/academy/dtoapp/model/Book.java (hunk 3 above)
Comment (artemklishch) [reply src/main/java/mate/academy/dtoapp/model/Book.java:23]: ++, added restriction to null value

File: src/main/java/mate/academy/dtoapp/model/Book.java (hunk 2 above)
Comment (artemklishch) [reply src/main/java/mate/academy/dtoapp/model/Book.java]: ++

File: src/main/java/mate/academy/dtoapp/controller/BookController.java (hunk 1 above)
Comment (artemklishch) [reply src/main/java/mate/academy/dtoapp/controller/BookController.java]: ++

**Comment by artemklishch** [comment]
Thank you! Updated!

пт, 5 июл. 2024 г. в 15:25, Olena Bruiako ***@***.***>:

File: src/main/java/mate/academy/dtoapp/model/Book.java (hunk 3 above)
Comment (Elena-Bruyako) [reply src/main/java/mate/academy/dtoapp/model/Book.java:23]: What about other fields?

File: src/main/java/mate/academy/dtoapp/repository/BookRepositoryImpl.java [hunk 4]
```diff
@@ -1,4 +1,48 @@
...
+    public Book createBook(Book book) {
+        EntityTransaction transaction = null;
+        try (EntityManager entityManager = entityManagerFactory.createEntityManager()) {
+            transaction = entityManager.getTransaction();
+            transaction.begin();
+            entityManager.persist(book);
+            transaction.commit();
+            return book;
+        } catch (RuntimeException e) {
+            if (transaction != null) {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/dtoapp/repository/BookRepositoryImpl.java:49]: rollback doesn't work with try with resources

File: src/main/java/mate/academy/dtoapp/repository/BookRepositoryImpl.java [hunk 5]
```diff
@@ -1,4 +1,48 @@
...
+                transaction.rollback();
+            }
+            throw new RuntimeException("Failed to create the book", e);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/dtoapp/repository/BookRepositoryImpl.java]: create custom exception

File: src/main/java/mate/academy/dtoapp/repository/BookRepositoryImpl.java [hunk 6]
```diff
@@ -1,4 +1,48 @@
...
+        try (EntityManager entityManager = entityManagerFactory.createEntityManager()) {
+            return entityManager.createQuery("SELECT e FROM Book e", Book.class).getResultList();
+        }
+    }
+
+    @Override
+    public Optional<Book> findById(Long id) {
+        try (EntityManager entityManager = entityManagerFactory.createEntityManager()) {
+            Book book = entityManager.find(Book.class, id);
+            return Optional.ofNullable(book);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/dtoapp/repository/BookRepositoryImpl.java]: Should we throw an exception if book not exists?

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Check comments

File: src/main/java/mate/academy/dtoapp/repository/BookRepositoryImpl.java (hunk 5 above)
Comment (artemklishch) [reply src/main/java/mate/academy/dtoapp/repository/BookRepositoryImpl.java]: ++, thank you! updated

File: src/main/java/mate/academy/dtoapp/repository/BookRepositoryImpl.java (hunk 4 above)
Comment (artemklishch) [reply src/main/java/mate/academy/dtoapp/repository/BookRepositoryImpl.java:49]: ++, thank you! removed.

File: src/main/java/mate/academy/dtoapp/model/Book.java (hunk 3 above)
Comment (artemklishch) [reply src/main/java/mate/academy/dtoapp/model/Book.java:23]: ++, thank you! updated

File: src/main/java/mate/academy/dtoapp/repository/BookRepositoryImpl.java (hunk 6 above)
Comment (artemklishch) [reply src/main/java/mate/academy/dtoapp/repository/BookRepositoryImpl.java]: ++, thank you! changed

File: src/main/java/mate/academy/dtoapp/repository/BookRepositoryImpl.java [hunk 7]
```diff
@@ -1,4 +1,50 @@
...
+            return entityManager.createQuery("SELECT e FROM Book e", Book.class).getResultList();
+        }
+    }
+
+    @Override
+    public Book findById(Long id) {
+        try (EntityManager entityManager = entityManagerFactory.createEntityManager()) {
+            return Optional.ofNullable(entityManager.find(Book.class, id))
+                    .orElseThrow(
+                            () -> new EntityNotFoundException("Book with id " + id + " not found")
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/dtoapp/repository/BookRepositoryImpl.java]:
```suggestion
                            () -> new DataProcessingException("Book with id " + id + " not found")
```

File: src/main/java/mate/academy/dtoapp/repository/BookRepositoryImpl.java [hunk 8]
```diff
@@ -1,4 +1,50 @@
...
+                    );
+        }
+    }
+
+    @Override
+    public Book createBook(Book book) {
+        try (EntityManager entityManager = entityManagerFactory.createEntityManager()) {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/dtoapp/repository/BookRepositoryImpl.java]: Don't use try with resources with transactions
add rollback

File: src/main/java/mate/academy/dtoapp/repository/BookRepositoryImpl.java [hunk 9]
```diff
@@ -1,4 +1,50 @@
...
+            EntityTransaction transaction = entityManager.getTransaction();
+            transaction.begin();
+            entityManager.persist(book);
+            transaction.commit();
+            return book;
+        } catch (RuntimeException e) {
+            throw new FailedToSaveDataException("Failed to create the book. Error: " + e);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/dtoapp/repository/BookRepositoryImpl.java]:
```suggestion
            throw new DataProcessingException("Failed to create the book. Error: " + e);
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]

File: src/main/java/mate/academy/dtoapp/repository/BookRepositoryImpl.java (hunk 8 above)
Comment (artemklishch) [reply src/main/java/mate/academy/dtoapp/repository/BookRepositoryImpl.java]: ++, thank you. updated.

File: src/main/java/mate/academy/dtoapp/repository/BookRepositoryImpl.java (hunk 7 above)
Comment (artemklishch) [reply src/main/java/mate/academy/dtoapp/repository/BookRepositoryImpl.java]: ++, thanks. updated

File: src/main/java/mate/academy/dtoapp/repository/BookRepositoryImpl.java (hunk 9 above)
Comment (artemklishch) [reply src/main/java/mate/academy/dtoapp/repository/BookRepositoryImpl.java]: ++, thanks. updated

**Review by Elena-Bruyako — APPROVED** [review]

**Pull request merged**